File: day02.lua
local part1,part2=0,0
function omit(t,skip)
    local tt={}
    for i,x in ipairs(t) do if i~=skip then table.insert(tt,x) end end
    return tt
end
function ok(tab,skip)
    t=omit(tab,skip)
    for i=1,#t-1 do
        if (t[i+1]-t[i])*(t[2]-t[1])<0 then return false end
        local e=math.abs(t[i+1]-t[i]);if e>3 or e==0 then return false end
    end
    return true
end
for x in io.lines("data/input02.txt") do
    local t={}
    for d in string.gmatch(x, "[^% ]+") do
        table.insert(t,tonumber(d))
    end
    if ok(t) then part1=part1+1 end
    for skip=1,#t do
        if ok(t,skip) then part2=part2+1; break; end
    end
end
print(part1,part2)

File: day02.c
#include <stdio.h>
int remap(int i,int skip){return i>=skip?i+1:i;}

int main(void){
    int part1=0, part2=0; char line[100]; int L[8];
    FILE *f=fopen("data/input02.txt","r");
    while(fgets(line,100,f)){
        int N=sscanf(line,"%d %d %d %d %d %d %d %d",L,L+1,L+2,L+3,L+4,L+5,L+6,L+7);
        int ok(int skip){
            for(int i=0; i<(skip==999?N-1:N-2); i++){
                int e1=L[remap(1,skip)]-L[remap(0,skip)];
                int e2=L[remap(i+1,skip)]-L[remap(i,skip)];
                if(e1*e2<=0 || e2*e2>9){ return 0; }
            }
            return 1;
        }
        part1+=ok(999);
        for(int skip=0; skip<N; skip++){
            if(ok(skip)){part2++; break;}
        }
    }
    printf("%d %d\n",part1,part2);
}